exercises/practice/rational-numbers/test/Tests.hs

```diff
@@ -0,0 +1,305 @@
+{-# LANGUAGE RecordWildCards #-}
+
+import Prelude hiding    (abs, div, Rational)
+import qualified Prelude as P
+import Data.Function     (on)
+import Data.Foldable     (for_)
+import Test.Hspec        (Spec, describe, it, shouldBe)
+import Test.Hspec.Runner (configFailFast, defaultConfig, hspecWith)
+
+import RationalNumbers
+  ( Rational
+  , abs
+  , reduce
+  , add
+  , sub
+  , mul
+  , div
+  , exprational
+  , expreal
+  , rational
+  )
+
+main :: IO ()
+main = hspecWith defaultConfig {configFailFast = True} specs
+
+specs :: Spec
+specs = do
+          describe "abs"         $ for_ absCases         $ testA abs
+          describe "reduce"      $ for_ reduceCases      $ testA reduce
+          describe "add"         $ for_ addCases         $ testB add
+          describe "sub"         $ for_ subCases         $ testB sub
+          describe "mul"         $ for_ mulCases         $ testB mul
+          describe "div"         $ for_ divCases         $ testB div
+          describe "exprational" $ for_ exprationalCases $ testC exprational
+          describe "expreal"     $ for_ exprealCases     $ testD expreal
+ where
+  testA f CaseA{..} = it descriptionA $ f (rational number1A)
+                                        `shouldBe` rational expectedA
+  testB f CaseB{..} = it descriptionB $ f (rational number1B) (rational number2B)
+                                        `shouldBe` rational expectedB
+  testC f CaseC{..} = it descriptionC $ f (rational number1C) number2C
+                                        `shouldBe` rational expectedC
+  testD f CaseD{..} = it descriptionD $ f number1D (rational number2D)
+                                        `shouldBeAround` expectedD
+  shouldBeAround = shouldBe `on` roundTo 2
+
+  roundTo :: Int -> Float -> Float
+  roundTo n = (/ 10 ^ n) . (fromInteger :: Integer -> Float) . round . (* 10 ^ n)
+
+data CaseA = CaseA { descriptionA :: String
+                   , number1A     :: (Integer, Integer)
+                   , expectedA    :: (Integer, Integer)
+                   }
+
+data CaseB = CaseB { descriptionB :: String
+                   , number1B     :: (Integer, Integer)
+                   , number2B     :: (Integer, Integer)
+                   , expectedB    :: (Integer, Integer)
+                   }
+
+data CaseC = CaseC { descriptionC :: String
+                   , number1C     :: (Integer, Integer)
+                   , number2C     :: Integer
+                   , expectedC    :: (Integer, Integer)
+                   }
+
+data CaseD = CaseD { descriptionD :: String
+                   , number1D     :: Float
+                   , number2D     :: (Integer, Integer)
+                   , expectedD    :: Float
+                   }
```

Comment:
These four data types _could_ be combined into two (`CaseUnary`, `CaseBinary a b c`). Is the separation into four cases deliberate?

Reply:
No it isn't, I just didn't pay attention to the economics of test cases.